File: kiem_tra_c/test_xep.c
/* Kiem chung phan XEP BAI va quy uoc DO KHOANG CACH cua the xep 2D.
 *
 * Khong mo cua so: chi goi thang cac ham tinh toan.
 * Day la ban chuyen sang C cua kiem_tra/test_xep.py, giu nguyen tung phep thu.
 */
#include "../loi_c/thu_vien_moi_noi.h"
#include "../loi_c/xep_2d.h"
#include "../loi_c/loi_chung.h"
#include <stdio.h>
#include <string.h>
#include <math.h>

static int so_loi = 0;

static void ktra(const char *ten, int dat, const char *chi_tiet)
{
    printf("  [%s] %s", dat ? "DAT" : "SAI", ten);
    if (chi_tiet && chi_tiet[0]) printf("   %s", chi_tiet);
    printf("\n");
    if (!dat) so_loi++;
}

static const KieuGhep *theo_ma(const char *ma)
{
    const KieuGhep *k = kieu_theo_ma(ma);
    if (!k) { printf("  [SAI] khong co kieu '%s'\n", ma); so_loi++; }
    return k;
}

/* Sinh mot duong cat, bao loi va thoat neu that bai */
static int sinh_thu(const char *ma, double duong_kinh, const double *tham_so,
                double x, DuongCat *ra)
{
    const KieuGhep *k = theo_ma(ma);
    GiaTriThamSo gt;
    char loi[CO_LOI] = "";
    int i;
    if (!k) return -1;
    memset(&gt, 0, sizeof(gt));
    for (i = 0; i < k->so_tham_so; i++) gt.gt[i] = tham_so[i];
    duong_cat_khoi_tao(ra);
    if (kieu_sinh(k, duong_kinh, &gt, x, ra, loi) != 0) {
        printf("  [SAI] sinh %s that bai: %s\n", ma, loi);
        so_loi++;
        return -1;
    }
    return 0;
}

int main(void)
{
    DuongCat d, d0, d90;
    MucBai muc[3];
    KetQuaXep kq, kq2;
    Xep2D *xep;
    KhungNhatCat khung[3];
    char loi[CO_LOI] = "", chi_tiet[128];
    double dai_khuc = 200.0, khe = 8.0;
    int i;

    /* Tham so cua tung kieu, theo dung thu tu khai bao trong thu vien */
    static const double TS_GOC_90[1]   = { 0.0 };            /* a */
    static const double TS_GOC_45[1]   = { 0.0 };            /* a */
    static const double TS_NHANH_T[3]  = { 60.0, 0.0, 0.0 }; /* d_chinh, khe_ho, a */
    static const char *MA[3] = { "goc_90", "goc_45", "nhanh_t_90" };
    static const double *TS[3] = { TS_GOC_90, TS_GOC_45, TS_NHANH_T };

    printf("=== 1. THU VIEN chi con dung 3 kieu ghep ===\n");
    snprintf(chi_tiet, sizeof(chi_tiet), "%s, %s, %s",
             THU_VIEN[0].ma, THU_VIEN[1].ma, THU_VIEN[2].ma);
    ktra("co dung 3 kieu", SO_KIEU_GHEP == 3, chi_tiet);

    if (sinh_thu("goc_90", 60.0, TS_GOC_90, 100.0, &d) == 0) {
        ktra("goc 90 do -> moi dau vat 45 do", strstr(d.ten, "45") != NULL, d.ten);
        duong_cat_giai_phong(&d);
    }
    if (sinh_thu("goc_45", 60.0, TS_GOC_45, 100.0, &d) == 0) {
        ktra("goc 45 do -> moi dau vat 22,5 do", strstr(d.ten, "22.5") != NULL, d.ten);
        duong_cat_giai_phong(&d);
    }
    if (sinh_thu("nhanh_t_90", 60.0, TS_NHANH_T, 100.0, &d) == 0) {
        ktra("nhanh chu T -> long yen ngua 90 do",
             strstr(d.ten, "Yen ngua") != NULL, d.ten);
        duong_cat_giai_phong(&d);
    }

    printf("\n=== 2. Goc ghep = 2 lan goc vat (up hai mat vat vao nhau) ===\n");
    for (i = 0; i < 2; i++) {
        double r = 30.0, goc_ghep = i == 0 ? 90.0 : 45.0;
        double x1, x2, goc_vat;
        if (sinh_thu(MA[i], 2 * r, TS[i], 100.0, &d) != 0) continue;
        duong_cat_pham_vi_x(&d, &x1, &x2);
        /* Be ngang mieng vat theo truc X = D * tan(goc_vat)
         * => goc = atan(rong / D) */
        goc_vat = atan((x2 - x1) / (2 * r)) * 180.0 / PI;
        snprintf(chi_tiet, sizeof(chi_tiet), "%.4f do", goc_vat);
        {
            char ten[96];
            snprintf(ten, sizeof(ten), "%s: goc vat do duoc = %g do",
                     MA[i], goc_ghep / 2);
            ktra(ten, fabs(goc_vat - goc_ghep / 2) < 1e-6, chi_tiet);
        }
        duong_cat_giai_phong(&d);
    }

    printf("\n=== 3. Goc dat mieng cat xoay ca duong cat ===\n");
    {
        double ts0[1] = { 0.0 }, ts90[1] = { 90.0 };
        int lech_dung = 1, x_khong_doi = 1;
        if (sinh_thu("goc_90", 60.0, ts0, 100.0, &d0) == 0 &&
            sinh_thu("goc_90", 60.0, ts90, 100.0, &d90) == 0) {
            if (d0.so_diem != d90.so_diem) {
                lech_dung = x_khong_doi = 0;
            } else {
                for (i = 0; i < d0.so_diem; i++) {
                    if (fabs((d90.diem[i].a - d0.diem[i].a) - 90.0) > 1e-9)
                        lech_dung = 0;
                    if (fabs(d0.diem[i].x - d90.diem[i].x) > 1e-12)
                        x_khong_doi = 0;
                }
            }
            ktra("xoay 90 do thi moi diem lech dung 90 do", lech_dung, "");
            ktra("xoay khong lam doi vi tri truc X", x_khong_doi, "");
            duong_cat_giai_phong(&d0);
            duong_cat_giai_phong(&d90);
        }
    }

    printf("\n=== 4. XEP BAI: khuc noi tiep nhau, cach dung khoang khe ===\n");
    for (i = 0; i < 3; i++) {
        const KieuGhep *k = theo_ma(MA[i]);
        double x;
        int j;
        if (!k) return 1;
        if (vi_tri_ke_tiep(60.0, muc, i, dai_khuc, khe, 20.0, &x, loi) != 0) {
            printf("  [SAI] vi_tri_ke_tiep: %s\n", loi);
            return 1;
        }
        snprintf(muc[i].ma, sizeof(muc[i].ma), "%s", MA[i]);
        memset(&muc[i].gia_tri, 0, sizeof(muc[i].gia_tri));
        for (j = 0; j < k->so_tham_so; j++) muc[i].gia_tri.gt[j] = TS[i][j];
        muc[i].x = x;
    }
    ket_qua_xep_khoi_tao(&kq);
    if (xep_bai(60.0, muc, 3, 0.0, &kq, loi) != 0) {
        printf("  [SAI] xep_bai: %s\n", loi);
        return 1;
    }
    snprintf(chi_tiet, sizeof(chi_tiet), "%.3f mm", muc[0].x);
    ktra("nhat cat dau: chua dau 20 + khuc 200 = 220",
         fabs(muc[0].x - 220.0) < 1e-9, chi_tiet);
    for (i = 1; i < 3; i++) {
        double x1, x_cuoi_truoc, dai;
        char ten[128];
        khung_duong_cat(&kq.duong[i - 1], &x1, &x_cuoi_truoc, &dai);
        snprintf(ten, sizeof(ten),
                 "nhat %d cach cho sau nhat cua nhat %d dung %g+%g mm",
                 i + 1, i, khe, dai_khuc);
        snprintf(chi_tiet, sizeof(chi_tiet), "%.1f mm", muc[i].x - x_cuoi_truoc);
        ktra(ten, fabs(muc[i].x - (x_cuoi_truoc + khe + dai_khuc)) < 1e-9, chi_tiet);
    }

    printf("\n=== 5. KHUNG duong cat = be ngang theo truc X ===\n");
    for (i = 0; i < kq.so_duong; i++) {
        double x1, x2, dai, nho, lon, thuc;
        char ten[96];
        int j;
        khung_duong_cat(&kq.duong[i], &x1, &x2, &dai);
        nho = lon = kq.duong[i].diem[0].x;
        for (j = 0; j < kq.duong[i].so_diem; j++) {
            double v = kq.duong[i].diem[j].x;
            if (v < nho) nho = v;
            if (v > lon) lon = v;
        }
        thuc = lon - nho;
        snprintf(ten, sizeof(ten), "khung %d rong dung bang be ngang duong cat", i + 1);
        snprintf(chi_tiet, sizeof(chi_tiet), "%.3f mm", dai);
        ktra(ten, fabs(dai - thuc) < 1e-12, chi_tiet);
    }

    printf("\n=== 6. DO KHOANG CACH tu DIEM GOC (dau ong xa mam kep) ===\n");
    xep = xep2d_tao(NULL);
    for (i = 0; i < kq.so_duong; i++) {
        double x1, x2, dai;
        khung_duong_cat(&kq.duong[i], &x1, &x2, &dai);
        khung[i].x_dau = x1;
        khung[i].x_cuoi = x2;
        snprintf(khung[i].ten, sizeof(khung[i].ten), "%s", kq.duong[i].ten);
        khung[i].x_tam = muc[i].x;
    }
    xep2d_dat_du_lieu(xep, khung, kq.so_duong, 60.0, 1200.0);
    xep2d_dat_co_khung_hinh(xep, 800, 300);
    for (i = 0; i < kq.so_duong; i++) {
        double x1, x2, dai, kc;
        char ten[96];
        khung_duong_cat(&kq.duong[i], &x1, &x2, &dai);
        kc = xep2d_khoang_cach_tu_goc(xep, x2);
        snprintf(ten, sizeof(ten),
                 "nhat %d: do toi CANH GAN diem goc nhat (x_cuoi)", i + 1);
        snprintf(chi_tiet, sizeof(chi_tiet), "%.1f mm", kc);
        ktra(ten, fabs(kc - (1200.0 - x2)) < 1e-12, chi_tiet);
    }
    {
        int giam_dan = 1;
        for (i = 0; i + 1 < kq.so_duong; i++) {
            double a1, a2, b1, b2, dai;
            khung_duong_cat(&kq.duong[i], &a1, &a2, &dai);
            khung_duong_cat(&kq.duong[i + 1], &b1, &b2, &dai);
            if (!(xep2d_khoang_cach_tu_goc(xep, a2) >
                  xep2d_khoang_cach_tu_goc(xep, b2)))
                giam_dan = 0;
        }
        ktra("nhat cang xa mam kep thi khoang cach cang nho", giam_dan, "");
    }

    printf("\n=== 7. Go khoang cach vao o -> nhat cat nhay dung cho ===\n");
    {
        static const double MUON[3] = { 100.0, 350.0, 800.0 };
        int t;
        for (t = 0; t < 3; t++) {
            MucBai muc2[3];
            double x1, x2, dai, x_moi, kc;
            char ten[96];
            i = 1;
            khung_duong_cat(&kq.duong[i], &x1, &x2, &dai);
            x_moi = xep2d_tu_khoang_cach(xep, MUON[t], muc[i].x, x2);
            memcpy(muc2, muc, sizeof(muc2));
            muc2[i].x = x_moi;
            ket_qua_xep_khoi_tao(&kq2);
            if (xep_bai(60.0, muc2, 3, 0.0, &kq2, loi) != 0) {
                printf("  [SAI] xep_bai: %s\n", loi);
                so_loi++;
                continue;
            }
            khung_duong_cat(&kq2.duong[i], &x1, &x2, &dai);
            kc = xep2d_khoang_cach_tu_goc(xep, x2);
            snprintf(ten, sizeof(ten), "go %g mm -> do lai duoc dung %g mm",
                     MUON[t], MUON[t]);
            snprintf(chi_tiet, sizeof(chi_tiet), "%.4f mm", kc);
            ktra(ten, fabs(kc - MUON[t]) < 1e-9, chi_tiet);
            ket_qua_xep_giai_phong(&kq2);
        }
    }

    printf("\n=== 8. Doi pixel <-> mm cua the xep khop nhau ===\n");
    {
        static const double MM[3] = { 0.0, 250.0, 1200.0 };
        int t;
        for (t = 0; t < 3; t++) {
            char ten[64];
            snprintf(ten, sizeof(ten), "%g mm -> pixel -> mm", MM[t]);
            ktra(ten, fabs(xep2d_sang_mm(xep, xep2d_sang_pixel(xep, MM[t])) - MM[t])
                      < 1e-9, "");
        }
        xep2d_phong_to(xep, 3.0, 400);
        ktra("sau khi phong to van khop",
             fabs(xep2d_sang_mm(xep, xep2d_sang_pixel(xep, 600.0)) - 600.0) < 1e-9, "");
        ktra("phong to giu nguyen diem duoi con tro",
             fabs(xep2d_sang_pixel(xep, xep2d_sang_mm(xep, 400)) - 400) < 1e-9, "");
    }

    printf("\n=== 9. Bao loi khi cay ong khong du dai ===\n");
    ket_qua_xep_khoi_tao(&kq2);
    if (xep_bai(60.0, muc, 3, 300.0, &kq2, loi) == 0) {
        ktra("bao THIEU khi cay ong ngan", strstr(kq2.canh_bao, "THIEU") != NULL,
             kq2.canh_bao[0] ? kq2.canh_bao : "khong bao gi");
        ket_qua_xep_giai_phong(&kq2);
    } else {
        printf("  [SAI] xep_bai (cay ngan): %s\n", loi);
        so_loi++;
    }
    ket_qua_xep_khoi_tao(&kq2);
    if (xep_bai(60.0, muc, 3, 2000.0, &kq2, loi) == 0) {
        ktra("khong bao gi khi cay ong du dai", kq2.canh_bao[0] == '\0', "");
        ket_qua_xep_giai_phong(&kq2);
    }

    ket_qua_xep_giai_phong(&kq);
    xep2d_giai_phong(xep);

    if (so_loi == 0) printf("\n=== TAT CA DAT ===\n");
    else             printf("\n=== CO %d LOI ===\n", so_loi);
    return so_loi ? 1 : 0;
}

File: kiem_tra/test_xep.py
"""Kiem chung phan XEP BAI va quy uoc DO KHOANG CACH cua the xep 2D.

Khong mo cua so: chi goi thang cac ham tinh toan.
"""
import math
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))
from loi import thu_vien_moi_noi as tv
from loi.xep_2d import Xep2D


class CanvasGia:
    """Thay cho Canvas that: bao kich thuoc, con moi lenh ve thi nuot khong lam gi."""
    def winfo_width(self):
        return 800

    def winfo_height(self):
        return 300

    def __getattr__(self, ten):        # delete / create_line / create_text ...
        return lambda *a, **k: None


loi = 0


def ktra(ten, dat, chi_tiet=""):
    global loi
    print(f"  [{'DAT' if dat else 'SAI'}] {ten}" + (f"   {chi_tiet}" if chi_tiet else ""))
    if not dat:
        loi += 1


print("=== 1. THU VIEN chi con dung 3 kieu ghep ===")
ktra("co dung 3 kieu", len(tv.THU_VIEN) == 3,
     ", ".join(k.ma for k in tv.THU_VIEN))
ktra("goc 90 do -> moi dau vat 45 do",
     "45" in tv.THEO_MA["goc_90"].sinh(60.0, {"a": 0.0}, 100.0).ten)
ktra("goc 45 do -> moi dau vat 22,5 do",
     "22.5" in tv.THEO_MA["goc_45"].sinh(60.0, {"a": 0.0}, 100.0).ten)
ktra("nhanh chu T -> long yen ngua 90 do",
     "Yen ngua" in tv.THEO_MA["nhanh_t_90"].sinh(
         60.0, {"d_chinh": 60.0, "khe_ho": 0.0, "a": 0.0}, 100.0).ten)

print("\n=== 2. Goc ghep = 2 lan goc vat (up hai mat vat vao nhau) ===")
for ma, goc_ghep in (("goc_90", 90.0), ("goc_45", 45.0)):
    r = 30.0
    d = tv.THEO_MA[ma].sinh(2 * r, {"a": 0.0}, 100.0)
    x1, x2 = d.pham_vi_x()
    # Be ngang mieng vat theo truc X = D * tan(goc_vat)  =>  goc = atan(rong / D)
    goc_vat = math.degrees(math.atan((x2 - x1) / (2 * r)))
    ktra(f"{ma}: goc vat do duoc = {goc_ghep / 2:g} do",
         abs(goc_vat - goc_ghep / 2) < 1e-6, f"{goc_vat:.4f} do")

print("\n=== 3. Goc dat mieng cat xoay ca duong cat ===")
d0 = tv.THEO_MA["goc_90"].sinh(60.0, {"a": 0.0}, 100.0)
d90 = tv.THEO_MA["goc_90"].sinh(60.0, {"a": 90.0}, 100.0)
ktra("xoay 90 do thi moi diem lech dung 90 do",
     all(abs((b[1] - a[1]) - 90.0) < 1e-9 for a, b in zip(d0.diem, d90.diem)))
ktra("xoay khong lam doi vi tri truc X",
     all(abs(a[0] - b[0]) < 1e-12 for a, b in zip(d0.diem, d90.diem)))

print("\n=== 4. XEP BAI: khuc noi tiep nhau, cach dung khoang khe ===")
muc, dai_khuc, khe = [], 200.0, 8.0
for ma, g in (("goc_90", {"a": 0.0}), ("goc_45", {"a": 0.0}),
              ("nhanh_t_90", {"d_chinh": 60.0, "khe_ho": 0.0, "a": 0.0})):
    x = tv.vi_tri_ke_tiep(60.0, muc, ma, g, dai_khuc, khe, chua_dau=20.0)
    muc.append({"ma": ma, "gia_tri": g, "x": x})
kq = tv.xep_bai(60.0, muc)
ktra("nhat cat dau: chua dau 20 + khuc 200 = 220",
     abs(muc[0]["x"] - 220.0) < 1e-9, f"{muc[0]['x']:.3f} mm")
for i in range(1, len(muc)):
    x_cuoi_truoc = tv.khung_duong_cat(kq.cac_duong[i - 1])[1]
    ktra(f"nhat {i + 1} cach cho sau nhat cua nhat {i} dung {khe:g}+{dai_khuc:g} mm",
         abs(muc[i]["x"] - (x_cuoi_truoc + khe + dai_khuc)) < 1e-9,
         f"{muc[i]['x'] - x_cuoi_truoc:.1f} mm")

print("\n=== 5. KHUNG duong cat = be ngang theo truc X ===")
for i, d in enumerate(kq.cac_duong):
    x1, x2, dai = tv.khung_duong_cat(d)
    thuc = max(p[0] for p in d.diem) - min(p[0] for p in d.diem)
    ktra(f"khung {i + 1} rong dung bang be ngang duong cat", abs(dai - thuc) < 1e-12,
         f"{dai:.3f} mm")

print("\n=== 6. DO KHOANG CACH tu DIEM GOC (dau ong xa mam kep) ===")
xep = Xep2D(CanvasGia())
xep.dat_du_lieu([(tv.khung_duong_cat(d)[0], tv.khung_duong_cat(d)[1], d.ten, m["x"])
                 for d, m in zip(kq.cac_duong, muc)], 60.0, 1200.0)
for i, d in enumerate(kq.cac_duong):
    x1, x2, _ = tv.khung_duong_cat(d)
    kc = xep.khoang_cach_tu_goc(x1, x2)
    ktra(f"nhat {i + 1}: do toi CANH GAN diem goc nhat (x_cuoi)",
         abs(kc - (1200.0 - x2)) < 1e-12, f"{kc:.1f} mm")
ktra("nhat cang xa mam kep thi khoang cach cang nho",
     all(xep.khoang_cach_tu_goc(*tv.khung_duong_cat(kq.cac_duong[i])[:2]) >
         xep.khoang_cach_tu_goc(*tv.khung_duong_cat(kq.cac_duong[i + 1])[:2])
         for i in range(len(kq.cac_duong) - 1)))

print("\n=== 7. Go khoang cach vao o -> nhat cat nhay dung cho ===")
for muon in (100.0, 350.0, 800.0):
    i = 1
    x1, x2, _ = tv.khung_duong_cat(kq.cac_duong[i])
    x_moi = xep.tu_khoang_cach(muon, x2 - x1, muc[i]["x"], x2)
    muc2 = [dict(m) for m in muc]
    muc2[i]["x"] = x_moi
    d_moi = tv.xep_bai(60.0, muc2).cac_duong[i]
    kc = xep.khoang_cach_tu_goc(*tv.khung_duong_cat(d_moi)[:2])
    ktra(f"go {muon:g} mm -> do lai duoc dung {muon:g} mm",
         abs(kc - muon) < 1e-9, f"{kc:.4f} mm")

print("\n=== 8. Doi pixel <-> mm cua the xep khop nhau ===")
for mm in (0.0, 250.0, 1200.0):
    ktra(f"{mm:g} mm -> pixel -> mm", abs(xep.sang_mm(xep.sang_pixel(mm)) - mm) < 1e-9)
xep.phong_to(3.0, 400)
ktra("sau khi phong to van khop", abs(xep.sang_mm(xep.sang_pixel(600.0)) - 600.0) < 1e-9)
ktra("phong to giu nguyen diem duoi con tro",
     abs(xep.sang_pixel(xep.sang_mm(400)) - 400) < 1e-9)

print("\n=== 9. Bao loi khi cay ong khong du dai ===")
kq2 = tv.xep_bai(60.0, muc, dai_cay_ong=300.0)
ktra("bao THIEU khi cay ong ngan", any("THIEU" in c for c in kq2.canh_bao),
     kq2.canh_bao[0] if kq2.canh_bao else "khong bao gi")
ktra("khong bao gi khi cay ong du dai",
     not tv.xep_bai(60.0, muc, dai_cay_ong=2000.0).canh_bao)

print(f"\n{'=== TAT CA DAT ===' if not loi else f'=== CO {loi} LOI ==='}")
sys.exit(1 if loi else 0)
